README.md
# js-top-down-space-shooter
This is a vanilla JavaScript canvas based scrolling top down space shooter.  That was a mouth full.

Live version of the game can be found at www.mhzsys.net/games/TDSS

Use the mouse to move the ship arround and click to fire the weapon.

## Known bugs
- [x] Weapon deletion sometimes results in error unable to read property of x, does not effect gameplay.
- [x] Weapons fire bounces off walls.
- [/] Ensure the click is only registered when inside play area.
- [ ] Center canvas in middle of page instead left aligned.
- [x] Fix json loading to object instead of array.

changelog.md
# Things to add or improve.
- [X] Create different types of enemy with unique aspects.
- [X] Make asteroid enemys that break into smaller asteroids.
- [X] Give enemy different ammounts of HP.
- [X] Change weapon damage.
- [ ] Add dynamic and changeable player ships.
- [X] Detect window size and adjust play window accordingly.
- [ ] More spread progression
- [ ] Create options menue in top right hand corner.
- [X] Add a space station at the bottom for asteroids to hit.
- [/] Give space station HP and when destroyed game over.
- [ ] Make red asteroids explode and destory other asteroids in AOE.
- [/] Add special boss asteroid to win the game.
- [ ] Refactor and optimize code for effecency.
- [ ] Add animated effects.
- [ ] Look for a way to check performace.
- [ ] Look into web workers for multi threading
- [X] Instead of looping BG, tile it to work with differnt height screens.
- [X] Add space bar for fireing becuse touch pads that are the click button suck.
- [ ] Refactor update and frame rate to be dynamic.
- [X] Center canvas in window.

# 2020-01-31
- Fixed json loading bug by converting it to an object instead of an array.
- Updated readme and jquery file.
- Moved VFX folder to assets and modified pointers to reflect.

# 2018-08-06
- Fixed background into a streaming bg instead of a rotating one.
- Added space bar to firing.
- Added final asteroid boss but does not spawn yet.
- Added space station with HP and a spawn for the final asteroid.

# 2018-08-05
- updated change log and things to add/improve section.

# 2018-08-04
- Added a scrolling stary background
- Added static asteroid images
- Set Asteroid VFX data to json file.
- Expanded FVX to lvl2 and lvl1 asteroids as well as randoms.
- Changed random asteroids to new grapic and gave red one to random lvl2.
- Weapons now have grapics and are refrenced in json file.

# 2018-08-03
- Added two more fireing modes, alternating fire, and cycle fire.
- Added new asteroid type enemy that breaks into smaller pieces after destruction.
- Player now takes damage equal to how much hp an enemy has left.
- Added test ultimate weapon combining 2 other weapons.
- Added another break to the lvl2 asteroids to lvl1 asteroids.
- Fixed asteroids bouncing off walls and now get removed when out of bounds.

# 2018-08-02
- Fixed mobile touch screen and weapons on mobile device.
- Moved enemy generation to its own enemys.js file.
- Weapons upgrade based on number of kills.
- New different colored enemy appears after 10 kills.
- Enemy now have their own HP.
- New organge enemy requires 2 hits to kill.
- weapons now have hp damage.

# 2018-08-01
- Prevent text select messing with canvas game.
- Added ability to change window size and canvas adjusts accordingly.
- Added touch screen control capability.

# 2018-07-31
- Added ability for multiple weapon modes selectable by the 1, 2 keys.  More to come.
- Added a dual fire mode set to key 3.
- Added spread fire mode to key 4.
- Added a not to great lazer beam weapon, needs work.
- Preventing weapons fire from bouncing off the walls.

# 2018-07-30
- When enemy hits player it is destroyed and only does 1 damage.
- For every 10 enemy shot and destroyed player regains 1 hp.
- Changed background to black.
- Changed projectile weapons color to silver.
- Hide mouse cursor when inside canvas element.
- Seperated controls to its own file.
- Seperated weapons to its own file.
- Disabled right click menue.
- Added a secondary fire mode as a new weapon constructor test.
- Fixed weapon destroying enemy error from console log resulting in screen flashing.

# 2018-07-29
- Created a top down space shooter based on the raining chain tutorial on youtube.  Current features include:
- 500 x 700 canvas play area.
- green square represents player1.
- player movement determined by mouse position.
- enemy are randomly generated and drop from top and fall to bottom.
- when enemy reaches bottom it is removed from play area.
- when player clicks a bullet is fired at current x position till it hits an enemy or reaches top of map.
- if bullet reaches top of map it is removed from the play area.
- if bullet hits an enemy block the enemy is destroyed and player score goes up by one.
- player has 10 HP and takes 1 hp of damage every frame an enemy is touching the player.


npc.js
'use strict';

let Npc = function (id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed) {
  let object = {
    id: id,
    name: name,
    x: x,
    y: y,
    vfx, vfx,
    spdX: spdX,
    spdY: spdY,
    width: width,
    height: height,
    color: color,
    hp: hp,
    afterDestroyed: afterDestroyed,
  };
  npcList[id] = object;
};

let npcVFX = function () {
  for (let key in npcList) {
    if (npcList[key].name === 'space station') {
      let vfx_npc = new Image();
      vfx_npc.src = "./assets/VFX/WB_base_d1.png";
      // console.log("space station hit")
      ctx.drawImage(
        vfx_npc,
        npcList[key].x - (npcList[key].width / 2),
        npcList[key].y - (npcList[key].height / 2),
        npcList[key].width,
        npcList[key].height
      )
    }
  }

}

let space_station = function () {
  console.log('space station generated')
  let id = Math.random();
  let name = 'space station'
  let x = canvasWidth / 2;
  let y = canvasHeight - 75;
  let vfx = null;
  let spdX = 0;
  let spdY = 0;
  let width = (canvasWidth / 3) * 2;
  let height = 150;
  let color = '#404040';
  let hp = 250;
  let afterDestroyed = 'lvl complete'

  Npc(id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed)
}

game.js
'use strict';

const ctx = document.getElementById('ctx').getContext('2d');
let windowHeight = window.innerHeight;
let windowWidth = window.innerWidth;
let canvasHeight = window.innerHeight - 6; // Y
let canvasWidth = 500; // X
setCanvas()
const jsonData = {};
let timeStarted = Date.now();
let background = {};
let enemyList = {};
let weaponsFire = {};
let npcList = {};
let frameCount = 0;
let score = 0;
let missed = 0;
let hpRegen = 0;
let firing = false;
let space = false;
let startedFiring = 0;
let fireRate = 0;
let weaponSelect = 0;
let isMobile = false;
let protecting = 0;
let paused = false;
let bg = new Image();
bg.src = "img/bg-stars-portrait-x500-y500.png";

// Detects if playing on a touch screen mobile device.
if (/Android|webOS|iPhone|iPad|BlackBerry|Windows Phone|Opera Mini|IEMobile|Mobile/i.test(navigator.userAgent)) {
  console.log('mobile deveice')
  isMobile = true;
} else {
  isMobile = false;
}

function setCanvas() {
  windowHeight = window.innerHeight;
  windowWidth = window.innerWidth;
  canvasHeight = windowHeight - 6;
  if (window.innerWidth > 500) {
    canvasWidth = 500 - 8;
  } else {
    canvasWidth = window.innerWidth;
  }
  if (window.innerHeight > 1000) {
    canvasHeight = 1000;
  } else {
    canvasHeight = window.innerHeight - 6;
  }
  document.getElementById('ctx').height = canvasHeight;
  document.getElementById('ctx').width = canvasWidth;
  ctx.font = '30px "Courier New", Courier, monospace';
  ctx.fillStyle = 'white';
}

window.onresize = function (event) {
  setCanvas();
};

let player1 = {
  name: 'P1',
  x: 250,
  y: 600,
  spdX: 25,
  spdY: 5,
  hp: 10,
  width: 40,
  height: 30,
  color: 'green',
};

let player1VFX = function () {
  let vfx_ship = new Image();
  vfx_ship.src = "./assets/VFX/ships/1B.png";
  ctx.drawImage(
    vfx_ship,
    player1.x - (player1.width / 2),
    player1.y - (player1.height / 2),
    player1.width,
    player1.height
  )
}

let updateEntity = function (update) {
  updateEntityPosition(update);
  drawEntity(update);
};

let updateEntityPosition = function (update) {
  update.x += update.spdX;
  update.y += update.spdY;
  // if (update.x < 0 || update.x > canvasWidth) {
  //   update.spdX = -update.spdX;
  // }
  // if (update.y < 0 || update.y > canvasHeight) {
  //   update.spdY = -update.spdX;
  // }
};

let testCollisionRectRect = function (rect1, rect2) {
  return rect1.x <= rect2.x + rect2.width
    && rect2.x <= rect1.x + rect1.width
    && rect1.y <= rect2.y + rect2.height
    && rect2.y <= rect1.y + rect1.height;
}

let drawEntity = function (draw) {
  ctx.save();
  // ctx.fillStyle = draw.color;
  // ctx.fillRect(draw.x - draw.width / 2, draw.y - draw.height / 2, draw.width, draw.height);
  npcVFX();
  player1VFX();
  enemyVFX();
  weaponVFX();

  ctx.restore();

};

let getDistanceBetweenEntity = function (entity1, entity2) {
  let vx = entity1.x - entity2.x;
  let vy = entity1.y - entity2.y;
  return Math.sqrt(vx * vx + vy * vy);
};

let collisionDetection = function (entity1, entity2) {
  let rect1 = {
    x: entity1.x - entity1.width / 2,
    y: entity1.y - entity1.height / 2,
    width: entity1.width,
    height: entity1.height,
  }
  let rect2 = {
    x: entity2.x - entity2.width / 2,
    y: entity2.y - entity2.height / 2,
    width: entity2.width,
    height: entity2.height,
  }
  return testCollisionRectRect(rect1, rect2);
};

let playerEnemyHitDetection = function () {
  for (let key in enemyList) {
    updateEntity(enemyList[key]);
    let isColliding = collisionDetection(player1, enemyList[key]);
    if (enemyList[key].x > canvasWidth || enemyList[key].x < 0) {
      delete enemyList[key];
    } else if (enemyList[key].y > canvasHeight) {
      missed++;
      delete enemyList[key];
    }
    if (isColliding) {
      player1.hp = player1.hp - enemyList[key].hp;
      delete enemyList[key];
    }
  }
  for (let key2 in npcList) {
    for (let key in enemyList) {
      let isColliding2 = collisionDetection(npcList[key2], enemyList[key]);
      if (isColliding2) {
        npcList[key2].hp = npcList[key2].hp - enemyList[key].hp;
        delete enemyList[key];
        console.log('space station hit: ', npcList[key2].hp)
      }
    }
  }
}

let playerWeaponHitDetection = function () {
  for (let key1 in enemyList) {
    for (let key2 in weaponsFire) {
      let isColliding2 = collisionDetection(weaponsFire[key2], enemyList[key1]);
      if (isColliding2) {
        enemyList[key1].hp = enemyList[key1].hp - weaponsFire[key2].damage
        delete weaponsFire[key2];
        if (enemyList[key1].hp < 1) {
          score++;
          hpRegen++;
          // console.log(enemyList[key1].afterDestroyed, enemyList[key1].x, enemyList[key1].y)
          afterEffect(enemyList[key1].afterDestroyed, enemyList[key1].x, enemyList[key1].y)
          delete enemyList[key1];
        }
        break;
      }
    }
  }
}

let strayBulletCleanup = function () {
  for (let key1 in weaponsFire) {
    updateEntity(weaponsFire[key1]);
    if (weaponsFire[key1].y <= 1 || weaponsFire[key1].x > canvasWidth || weaponsFire[key1].x < 0) {
      delete weaponsFire[key1];
    }
  }
}

let fireSelectedWeapon = function (mouse_X, mouse_Y) {
  if (weaponSelect == 0) fireWeapon(mouse_X, mouse_Y);
  if (weaponSelect == 1) fireWeapon2(mouse_X, mouse_Y);
  if (weaponSelect == 2) fireWeapon3(mouse_X, mouse_Y);
  if (weaponSelect == 3) fireWeapon4(mouse_X, mouse_Y);
  if (weaponSelect == 4) fireWeapon5(mouse_X, mouse_Y);
  if (weaponSelect == 5) fireWeapon6(mouse_X, mouse_Y);
  if (weaponSelect == 6) fireWeapon7(mouse_X, mouse_Y);
  if (weaponSelect == 7) fireWeapon8(mouse_X, mouse_Y);
}

let background_y = -2000


let Background = function (id, bg, x, y) {
  let back = {
    id: id,
    bg: bg,
    x: x,
    y: y,
  }
  background[id] = back;
}
// for (let i = 0; i < Math.ceil(canvasHeight / 500); i++){
//   Background(Math.random(), bg, 0, ((i + 500) * -1));
//   console.log('initial bg')
// }

Background(Math.random(), bg, 0, -500);
Background(Math.random(), bg, 0, 0);
Background(Math.random(), bg, 0, 500);
Background(Math.random(), bg, 0, 1000);

let drawBackground = function () {
  // debugger;
  if (background_y >= 0) {
    // console.log('background reset')
    // background_y = background_y
  } else {
    background_y++;
  }

  for (let key in background) {
    ctx.drawImage(bg, 0, background[key].y)
    background[key].y++;
    if (background[key].y === 0) {
      console.log('creating new background')
      Background(Math.random(), bg, 0, -500);
    } else if (background[key].y === canvasHeight + 500) {
      delete background[key];
    }
  }
  // ctx.drawImage(bg, 0, canvasHeight - 1500)
  // ctx.drawImage(bg, 0, canvasHeight - 1000)
  // ctx.drawImage(bg, 0, canvasHeight - 500)
}

/* ---------------------------update------------------------------ */
let update = function () {
  if (paused) {
    console.log('Paused: ', frameCount);
    return;
  };
  ctx.clearRect(0, 0, canvasWidth, canvasHeight);//clears old data
  frameCount++;
  drawBackground();

  if (frameCount % 150 === 0) {
    for (let i = 0; i <= Math.random() * 3; i++) {
      randomlyGenerateEnemy();
    }
    asteroid_lvl3();
  }

  if (frameCount === 2000) {
    asteroid_final()
  }

  // Weapon upgrades
  if (score == 10) weaponSelect = 1;
  if (score == 20) weaponSelect = 2;
  if (score == 30) weaponSelect = 3;
  if (score == 40) weaponSelect = 4;
  if (score == 50) weaponSelect = 5;
  if (score == 60) weaponSelect = 6;
  if (score == 70) weaponSelect = 7;

  playerEnemyHitDetection();
  playerWeaponHitDetection();
  strayBulletCleanup();

  // Controls fireing rate
  if (firing == true) {
    if (startedFiring == 1) {
      fireSelectedWeapon(mouse_X, mouse_Y);
    }
    if (startedFiring > fireRate) {
      startedFiring = 0;
    }
    startedFiring++;
  }

  if (hpRegen == 10) {
    player1.hp++;
    level_1_Enemy();
    hpRegen = 0;
  }

  if (player1.hp <= 0) {
    console.log('You killed ', score, ' enemys but missed ', missed);
    alert('You killed ' + score + ' enemys but missed ' + missed)
    startNewGame();
  }



  drawEntity(player1);
  ctx.fillText(player1.hp + " HP", 0, 30);
  ctx.fillText('score: ' + score, 200, 30);
};

let startNewGame = function () {
  timeStarted = Date.now();
  player1.hp = 10;
  frameCount = 0;
  score = 0;
  missed = 0
  enemyList = {};
  weaponsFire = {};
  weaponSelect = 0;
  if (frameCount == 50) {
    randomlyGenerateEnemy();
    randomlyGenerateEnemy();
    randomlyGenerateEnemy();
  }
}

window.onload = function () {// Prevents function call till after page has fully loaded.
  startNewGame();
  space_station()
};


setInterval(update, 40); //40ms is equivelint to 24fps

enemys.js
'use strict';

$.getJSON('./assets/json/asteroids-large.json', function (data) {
  // jsonData.push(data)
  jsonData.enemy = data
  // console.log(jsonData);
});

let Enemy = function (id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed) {
  let enemy = {
    id: id,
    name: name,
    x: x,
    y: y,
    vfx, vfx,
    spdX: spdX,
    spdY: spdY,
    width: width,
    height: height,
    color: color,
    hp: hp,
    afterDestroyed: afterDestroyed,
  };
  enemyList[id] = enemy;
};

// context.drawImage(img,x,y,width,height);
// context.drawImage(img,sx,sy,swidth,sheight,x,y,width,height);
let enemyVFX = function () {
  for (let key in enemyList) {
    if (enemyList[key].name === 'random') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = jsonData.enemy.type.LA03 + jsonData.enemy.LA03[enemyList[key].vfx] + ".png";
      // console.log("random hit")
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    } else if (enemyList[key].name === 'asteroid lvl3') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = jsonData.enemy.type.LA01 + jsonData.enemy.LA01[enemyList[key].vfx] + ".png";
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    } else if (enemyList[key].name === 'asteroid lvl2') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = jsonData.enemy.type.LA01 + jsonData.enemy.LA01[enemyList[key].vfx] + ".png";
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    } else if (enemyList[key].name === 'asteroid lvl1') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = jsonData.enemy.type.LA01 + jsonData.enemy.LA01[enemyList[key].vfx] + ".png";
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    } else if (enemyList[key].name === 'randomlvl2') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = jsonData.enemy.type.LA02 + jsonData.enemy.LA02[enemyList[key].vfx] + ".png";
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    } else if (enemyList[key].name === 'asteroid final') {
      let vfx_asteroids = new Image();
      vfx_asteroids.src = "./assets/VFX/asteroids/asteroid_boss.png";
      ctx.drawImage(
        vfx_asteroids,
        enemyList[key].x - (enemyList[key].width / 2),
        enemyList[key].y - (enemyList[key].height / 2),
        enemyList[key].width,
        enemyList[key].height
      )
    }
    // if (enemyList[key].vfx > 0) {
    //   enemyList[key].vfx++;
    //   if (enemyList[key].vfx > 15) enemyList[key].vfx = 1
    // } else if (enemyList[key].vfx < 0) {
    //   enemyList[key].vfx--;
    //   if (enemyList[key].vfx < -15) enemyList[key].vfx = -1
    // }
  }

}

function afterEffect(option, x, y) {
  if (option == 'asteroid_lvl2') {
    asteroid_lvl2(x + 10, y, 1)
    asteroid_lvl2(x - 10, y, -1)
  }
  if (option == 'asteroid_lvl1') {
    asteroid_lvl1(x + 10, y + 10, 1, 1)
    asteroid_lvl1(x - 10, y - 10, -1, -1)
    asteroid_lvl1(x + 10, y + 10, 1, -1)
    asteroid_lvl1(x - 10, y - 10, -1, 1)
  }
}

let randomlyGenerateEnemy = function () {
  let id = Math.random();
  let name = 'random'
  let x = Math.round(Math.random() * canvasWidth);
  let y = 0;
  let height = 20 + Math.round(Math.random() * 30);
  let width = 20 + Math.round(Math.random() * 30);
  let vfx = 1 + Math.round(Math.random() * jsonData.enemy.LA03[0].length);
  let spdX = 0;
  let spdY = 2 + Math.round(Math.random() * 2);
  let color = 'red';
  let hp = 1;

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp)
}

let level_1_Enemy = function () {
  let id = Math.random();
  let name = 'randomlvl2'
  let x = Math.round(Math.random() * canvasWidth);
  let y = 0;
  let height = 20 + Math.round(Math.random() * 20);
  let width = 20 + Math.round(Math.random() * 20);
  let vfx = 1 + Math.round(Math.random() * jsonData.enemy.LA01[0].length);
  let spdX = 0;
  let spdY = 2 + Math.round(Math.random() * 2);
  let color = 'orange';
  let hp = 2;

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp)
}

let asteroid_lvl1 = function (x, y, spdX, spdY) {
  let id = Math.random();
  let name = 'asteroid lvl1'
  // if(x == null)
  // let x = Math.round(Math.random() * canvasWidth);
  // if(y == null)
  // let y = 0;
  let height = 20;
  let width = 20;
  let vfx = 1 + Math.round(Math.random() * 2);
  // let spdX = 0;
  // let spdY = 2 + Math.round(Math.random() * 2);
  let color = '#b3b3b3';
  let hp = 1;

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp)
}

let asteroid_lvl2 = function (x, y, spdX) {
  let id = Math.random();
  let name = 'asteroid lvl2'
  // let x = Math.round(Math.random() * canvasWidth);
  // let y = 0;
  let height = 45;
  let width = 45;
  let vfx = 1 + Math.round(Math.random() * jsonData.enemy.LA01[0].length);
  // let spdX = 0;
  let spdY = 1 + Math.round(Math.random() * 2);
  let color = '#808080';
  let hp = 2;
  let afterDestroyed = 'asteroid_lvl1'

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed)
}

let asteroid_lvl3 = function () {
  let id = Math.random();
  let name = 'asteroid lvl3'
  let x = Math.round(Math.random() * canvasWidth);
  let y = 0;
  let height = 75;
  let width = 75;
  let vfx = 1 + Math.round(Math.random() * jsonData.enemy.LA01[0].length);
  let spdX = 0;
  let spdY = 2;
  let color = '#404040';
  let hp = 3;
  let afterDestroyed = 'asteroid_lvl2'

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed)
}

let asteroid_final = function () {
  let id = Math.random();
  let name = 'asteroid final'
  let x = canvasWidth / 2;
  let y = -150;
  let height = 300;
  let width = 200;
  let vfx = null;
  let spdX = 0;
  let spdY = .5;
  let color = '#404040';
  let hp = 200;
  let afterDestroyed = 'lvl complete'

  Enemy(id, name, x, y, vfx, spdX, spdY, width, height, color, hp, afterDestroyed)
}

controls.js
'use strict';

let mouse_X = {};
let mouse_Y = {};

ctx.onselectstart = function () { return false; }

document.oncontextmenu = function () { // Disables right click menue
  console.log('right click')
  return false;
}

if (isMobile === false) {
  document.onmousemove = function (mouse) {
    let mouseX = mouse.clientX - document.getElementById('ctx').getBoundingClientRect().left;
    let mouseY = mouse.clientY - document.getElementById('ctx').getBoundingClientRect().top;
    mouse_X = mouseX;
    mouse_Y = mouseY;

    /*Prevents player from leaving field of play*/
    //Left
    if (mouse_X < player1.width / 2)
      mouse_X = player1.width / 2;
    //Right
    if (mouse_X > canvasWidth - player1.width / 2)
      mouse_X = canvasWidth - player1.width / 2;
    //Top
    if (mouse_Y < player1.height / 2)
      mouse_Y = player1.height / 2;
    //Bottom
    if (mouse_Y > canvasHeight - player1.height / 2)
      mouse_Y = canvasHeight - player1.height / 2;

    player1.x = mouse_X
    player1.y = mouse_Y
    // console.log('x: ', mouseX, 'y: ', mouseY);
  }

  document.onmousedown = function () {
    if (mouse_X <= canvasWidth && mouse_Y <= canvasHeight) {
      firing = true;
    }
  }

  document.onmouseup = function (mouse) {
    firing = false;
    startedFiring = 0;
  }
}

function touchUpdate(e) {
  let mouseX = e.touches[0].clientX - document.getElementById('ctx').getBoundingClientRect().left;
  let mouseY = e.touches[0].clientY - document.getElementById('ctx').getBoundingClientRect().top;
  mouse_X = mouseX;
  mouse_Y = mouseY;

  player1.x = mouse_X;
  player1.y = mouse_Y - 50; // Offsets above where user is touching
}

window.addEventListener('touchstart', function (e) {
  firing = true;
  touchUpdate(e);
});

window.addEventListener('touchmove', function (e) {
  touchUpdate(e);
});

window.addEventListener('touchend', function (e) {
  firing = false;
  startedFiring = 0;
});

document.onkeydown = function (event) {
  if (event.keyCode === 68 || event.keyCode === 39) { //d or Right arrow
    playerBottom.pressingRight = true;
    playerTop.pressingRight = true;
  } else if (event.keyCode === 83 || event.keyCode === 40) { //s or Down arrow
    playerLeft.pressingDown = true;
    playerRight.pressingDown = true;
  } else if (event.keyCode === 65 || event.keyCode === 37) { //a or Left arrow
    playerBottom.pressingLeft = true;
    playerTop.pressingLeft = true;
  } else if (event.keyCode === 87 || event.keyCode === 38) { // w or Up Arrow
    playerLeft.pressingUp = true;
    playerRight.pressingUp = true;
  } else if (event.keyCode === 80) { //p
    paused = !paused;
  } else if (event.keyCode === 32) { //space
    firing = true;
  } else if (event.keyCode === 49) { // 1
    weaponSelect = 0;
  } else if (event.keyCode === 50) { // 2
    weaponSelect = 1;
  } else if (event.keyCode === 51) { // 3
    weaponSelect = 2;
  } else if (event.keyCode === 52) { // 4
    weaponSelect = 5;
  }
};

document.onkeyup = function (event) {
  if (event.keyCode === 68 || event.keyCode === 39) { //d or Right arrow
    playerBottom.pressingRight = false;
    playerTop.pressingRight = false;
  } else if (event.keyCode === 83 || event.keyCode === 40) { //s or Down arrow
    playerLeft.pressingDown = false;
    playerRight.pressingDown = false;
  } else if (event.keyCode === 65 || event.keyCode === 37) { //a or Left arrow
    playerBottom.pressingLeft = false;
    playerTop.pressingLeft = false;
  } else if (event.keyCode === 87 || event.keyCode === 38) { // w or Up Arrow
    playerLeft.pressingUp = false;
    playerRight.pressingUp = false;
  } else if (event.keyCode === 32) { //space
    firing = false;
  }
};

weapons.js
'use strict';

$.getJSON('./assets/json/weapons.json', function (data) {
  // jsonData.push(data)
  jsonData.weapons = data
  // console.log(jsonData);
});

let altFire = false;
let cycleFire = 0;

let Weapon = function (id, x, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage) {
  let weapon = {
    id:id,
    x:x,
    y:y,
    vfx:vfx,
    spdX:spdX,
    spdY:spdY,
    width:width,
    height:height,
    color:color,
    type:type,
    fireRate:fireRate,
    damage:damage,
  }
  weaponsFire[id] = weapon;
}

let weaponVFX = function () {
  for (let key in weaponsFire) {
    if (weaponsFire[key].type === 'projectile') {
      let vfx_weaponsFire = new Image();
      vfx_weaponsFire.src = jsonData.weapons.type.Pulse + jsonData.weapons.Pulse[weaponsFire[key].vfx];
      ctx.drawImage(
        vfx_weaponsFire,
        weaponsFire[key].x - (weaponsFire[key].width / 2),
        weaponsFire[key].y - (weaponsFire[key].height / 2),
        weaponsFire[key].width,
        weaponsFire[key].height
      )
    } else if (weaponsFire[key].type === 'beam') {
      let vfx_weaponsFire = new Image();
      vfx_weaponsFire.src = jsonData.weapons.type.Lazer + jsonData.weapons.Lazer[weaponsFire[key].vfx];
      ctx.drawImage(
        vfx_weaponsFire,
        weaponsFire[key].x - (weaponsFire[key].width / 2),
        weaponsFire[key].y - (weaponsFire[key].height / 2),
        weaponsFire[key].width,
        weaponsFire[key].height
      )
    }
  }
}

// Single shot.
let fireWeapon = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 0;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -7;
  let color = 'silver';
  let type = 'projectile';
  let damage = 1;

  fireRate = 10;

  if (isMobile === true){
    y = mouseY - 50;
  }

  // console.log('created: ', id, x, y, spdX, spdY, width, height, color, type);
  Weapon(id, x, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
}

// Single shot increaded rate of fire.
let fireWeapon2 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 0;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -14;
  let color = 'orange';
  let type = 'projectile';
  let damage = 1;

  fireRate = 5;

  if (isMobile === true){
    y = mouseY - 50;
  }

  // console.log('created: ', id, x, y, spdX, spdY, width, height, color, type);
  Weapon(id, x, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
}

// Dual Fire
let fireWeapon3 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 1;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -14;
  let color = 'green';
  let type = 'projectile';
  let damage = 1;

  fireRate = 10;

  if (isMobile === true){
    y = mouseY - 50;
  }

  // console.log('created: ', id, x, y, spdX, spdY, width, height, color, type);
  Weapon(id, x - 4, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
  Weapon(id + 1, x +3 , y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
}

// Alternating Fire
let fireWeapon4 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 1;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -14;
  let color = 'green';
  let type = 'projectile';
  let damage = 1;

  fireRate = 5;

  if (isMobile === true){
    y = mouseY - 50;
  }
  
  if (altFire === false){
    Weapon(id, x - 4, y, vfx,  spdX, spdY, width, height, color, type, fireRate, damage);
  } else {
    Weapon(id, x + 4, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
  }
  altFire = !altFire;
}

// Spread Fire
let fireWeapon5 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 2;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -14;
  let color = 'purple';
  let type = 'projectile';
  let damage = 1;

  fireRate = 10;

  if (isMobile === true){
    y = mouseY - 50;
  }

  // console.log('created: ', id, x, y, spdX, spdY, width, height, color, type);
  Weapon(id, x - 4, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
  Weapon(id + 1, x + 4 , y, vfx,  spdX, spdY, width, height, color, type, fireRate, damage);
  Weapon(id + 2, x - 3 , y, vfx,  spdX + 5, spdY, width, height, color, type, fireRate, damage);
  Weapon(id + 3, x + 3 , y, vfx,  spdX - 5, spdY, width, height, color, type, fireRate, damage);
}

// Cycling Spread Fire
let fireWeapon6 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 2;
  let height = 10;
  let width = 5;
  let spdX = 0;
  let spdY = -14;
  let color = 'purple';
  let type = 'projectile';
  let damage = 1;

  fireRate = 2;

  if (isMobile === true){
    y = mouseY - 50;
  }

  if (cycleFire == 0) Weapon(id, x - 4, y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
  if (cycleFire == 1) Weapon(id + 1, x + 4 , y, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
  if (cycleFire == 2) Weapon(id + 2, x - 3 , y, vfx, spdX + 5, spdY, width, height, color, type, fireRate, damage);
  if (cycleFire == 3) Weapon(id + 3, x + 3 , y, vfx, spdX - 5, spdY, width, height, color, type, fireRate, damage);
  
  if (cycleFire == 4){
    cycleFire = 0
  } else {
    cycleFire++;
  }
}

// Lazer Beam.
let fireWeapon7 = function (mouseX, mouseY) {
  let id = Math.random();
  let x = mouseX;
  let y = mouseY;
  let vfx = 8;
  let height = 150;
  let width = 5;
  let spdX = 0;
  let spdY = -20;
  let color = 'red';
  let type = 'beam';
  let damage = 3;

  fireRate = 25;

  if (isMobile === true){
    y = mouseY - 50;
  }

  // console.log('created: ', id, x, y, spdX, spdY, width, height, color, type);
  Weapon(id, x, y - 100, vfx, spdX, spdY, width, height, color, type, fireRate, damage);
}

// Lazer Beam and Projectiles
let fireWeapon8 = function(mouseX, mouseY) {
  fireWeapon7(mouseX, mouseY);
  fireWeapon6(mouseX, mouseY);
}